celo-org/wallet
# Fix for tapping on a push notification on android

### Description

This is a fix for tapping on a push notification. Currently it is not handled properly if the app was backgrounded via the back button on android or swiped away (killed state).

We were setting up firebase too early in the app lifecycle. The firebase messaging setup needs to happen in sync with the App component because intents on android travel through Activities. This is mentioned briefly in the documentation here: https://rnfirebase.io/messaging/notifications

`The call to getInitialNotification should happen within a React lifecycle method after mounting (e.g. componentDidMount or useEffect). If it's called too soon (e.g. within a class constructor or global scope), the notification data may not be available.`

### Tested

On an android emulator using the push script then tapping on the notification to see if the webview shows up.

Here are the different scenarios tested.
1. Start app, send push notice blue notification bar shows up at the top with the notification message.
2. Kill app (swipe away), send push, tap on it from the android status bar, make sure the app opens and navigates to the webview.
3. Open app, press android home button (app goes to background but main activity is paused not destroyed), send push, tap on it, make sure webview opens.
4. Open app, press android back button (app goes to background with main activity destroyed), send push, tap on it, make sure webview opens.

I also made sure there aren't any obvious leaks in either the channel use to listen for notifications or the callbacks by printing and pressing back and reopening the app to make sure there aren't multiple. But I am not entirely sure if the cleanup is properly done, so hopefully someone can confirm.

Added a new e2e test that simulates opening the app from a notification. It is currently disabled because circle ci does not support firebase yet. It runs successfully locally and hopefully soon @non-binary and @MuckT will help support firebase on CI.

### Related issues

- Fixes #161 

### Backwards compatibility

Yes.
